### PR TITLE
[Merged by Bors] - feat(RingTheory/PowerSeries/Basic): coeff of products

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
* Use the class `HasPiAntidiagonal` defined in PR #7904 to compute the coefficients of products of power series (in several or one variable) : `MvPowerSeries.coeff_prod` and `PowerSeries.coeff_prod`
* Update the file `Archive/Partition.lean` accordingly

Co-author : Maria Ines de Frutos Fernandez

Based on work of Bhavik Mehta in `Archive/Partition.lean`

---

- depends on: #7904 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/PowerSeries/Basic.lean [hunk 1]
```diff
@@ -812,6 +813,67 @@ set_option linter.uppercaseLean3 false in
...
+      rintro ⟨k, _, rfl⟩
+      rintro ⟨l, _, hkl⟩
+      simp only [Function.Embedding.coeFn_mk] at hkl
+      rw [FunLike.ext_iff] at hkl
+      specialize hkl a
+      simp only [Finsupp.coe_update, Function.update_same] at hkl
+      simp only [hkl.symm, ← huv', add_right_inj] at huv
+      apply h
+      simp only [Prod.mk.inj_iff]
+      exact ⟨hkl.symm, huv⟩
```
Comment (jcommelin) [inline Mathlib/RingTheory/PowerSeries/Basic.lean:873]:
```suggestion
  revert d
  induction s using Finset.induction_on with
  | empty =>
    intro d
    simp only [prod_empty, sum_const, nsmul_eq_mul, mul_one, coeff_one, piAntidiagonal_empty]
    split_ifs
    · simp only [card_singleton, Nat.cast_one]
    · simp only [card_empty, Nat.cast_zero]
  | @insert a s ha ih =>
    intro d
    rw [piAntidiagonal_insert ha, prod_insert ha, coeff_mul, sum_biUnion]
    · apply Finset.sum_congr rfl
      · simp only [mem_antidiagonal, sum_map, Function.Embedding.coeFn_mk, coe_update, Prod.forall]
        rintro u v rfl
        rw [ih, Finset.mul_sum, ← Finset.sum_attach]
        apply Finset.sum_congr rfl
        simp only [mem_attach, Finset.prod_insert ha, Function.update_same, forall_true_left,
          Subtype.forall]
        rintro x -
        rw [Finset.prod_congr rfl]
        intro i hi
        rw [Function.update_noteq]
        exact ne_of_mem_of_not_mem hi ha
    · simp only [Set.PairwiseDisjoint, Set.Pairwise, mem_coe, mem_antidiagonal, ne_eq,
        disjoint_left, mem_map, mem_attach, Function.Embedding.coeFn_mk, true_and, Subtype.exists,
        exists_prop, not_exists, not_and, forall_exists_index, and_imp, forall_apply_eq_imp_iff₂,
        Prod.forall, Prod.mk.injEq]
      rintro u v rfl u' v' huv h k - l - hkl
      obtain rfl : u' = u := by
        simpa only [Finsupp.coe_update, Function.update_same] using FunLike.congr_fun hkl a
      simp only [add_right_inj] at huv
      exact h rfl huv.symm
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Comment by Vierkantor** [comment]
Thanks for fixing the merge! Once the build succeeds, please reply with `bors r+`.

bors d+

**Pull request closed without merging**